### PR TITLE
feature(implement token on signup): provide user with token on signup

### PR DESCRIPTION
**Pull request by Billmike**
#### What does this PR do?
Enable the provision of token to users on signup
#### Description of Task to be completed?
- install lodash, chai and jsonwebtoken dependencies
- write custom validations for validating user input
- write unit test for signup features
- modify test script in package.json
- write controllers for client-side and server-side
- modify game.js file
#### How should this be manually tested?
After cloning to local machine, change directory into kenya-cfh and run `npm install` to install dependencies.
- run `npm run heroku-postbuild` to build the application
- run `npm test` to ensure all tests are passing
- run `npm start` to startup application and open browser at localhost:3000. Click on sign-up and fill in the user details
- To test with Postman, enter the url `localhost:3000/api/auth/signup` and enter the parameters `name`, `email` and `password`.
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
#153765403
#### Screenshots (if appropriate)
token

screen shot 2018-01-10 at 2 32 45 pm

error2

error1

#### Questions:
- N/A

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -36,153 +47,193 @@ exports.signup = function(req, res) {
...
-        }
-        if (!duplicate) {
-          console.log('Validated donation');
-          user.donations.push(req.body);
-          user.premium = 1;
-          user.save();
-        }
-      });
+          if (!duplicate) {
+            console.log('Validated donation');
```
Comment (temilaj) [inline app/controllers/users.js]: you should totally avoid logging to the console.

File: config/socket/game.js [hunk 2]
```diff
@@ -260,9 +260,10 @@ Game.prototype.shuffleCards = function(cards) {
 };
 
 Game.prototype.dealAnswers = function(maxAnswers) {
+  var self = this;
   maxAnswers = maxAnswers || 10;
   var storeAnswers = function(err, data) {
-    this.answers = data;
+    self.answers = data;
   };
   for (var i = 0; i < this.players.length; i++) {
```
Comment (temilaj) [inline config/socket/game.js]: Can we switch this from using an iterator to using an array method?

**Review by temilaj — CHANGES_REQUESTED** [review]

File: test/user/authentication.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        .end((err, res) => {
+          expect(res.statusCode).to.equal(400);
+          done();
+        });
+    }
+  );
+  it(
+    'Should return a 400 when a user tries to signup without a password',
+    (done) => {
+      const testUser = Object.assign({}, dummyUser);
```
Comment (temilaj) [inline test/user/authentication.js:63]: You should prefer the object spread operator over Object.assign to shallow-copy objects.

**Review by temilaj — CHANGES_REQUESTED** [review]

File: config/socket/game.js [hunk 4]
```diff
@@ -260,18 +260,19 @@ Game.prototype.shuffleCards = function(cards) {
...
+  var self = this;
   maxAnswers = maxAnswers || 10;
   var storeAnswers = function(err, data) {
-    this.answers = data;
+    self.answers = data;
   };
-  for (var i = 0; i < this.players.length; i++) {
-    while (this.players[i].hand.length < maxAnswers) {
-      this.players[i].hand.push(this.answers.pop());
+  this.players.map((player) => {
```
Comment (hay-bams) [inline config/socket/game.js:268]: Great job refactoring this to use map as opposed to a regular iterator

File: app/controllers/users.js [hunk 5]
```diff
@@ -36,153 +47,192 @@ exports.signup = function(req, res) {
...
+ * @returns {func} Returns the found user
  */
-exports.user = function(req, res, next, id) {
+exports.user = (req, res, next, id) => {
   User
     .findOne({
       _id: id
     })
-    .exec(function(err, user) {
+    .exec((err, user) => {
```
Comment (hay-bams) [inline app/controllers/users.js:232]: Great Job here, I can see you modified the code to use ES6 arrow function.

File: config/socket/game.js (hunk 4 above)
Comment (Billmike) [reply config/socket/game.js:268]: Thanks for the feedback

File: app/controllers/users.js (hunk 5 above)
Comment (Billmike) [reply app/controllers/users.js:232]: Thanks for the feedback.

**Pull request merged**